File: 567-permutation-in-string/567-permutation-in-string.java
class Solution {
    public boolean checkInclusion(String s1, String s2) {
        
        //1. create character freq map for the pattern
        Map<Character, Integer> freqMap = new HashMap<>();
        for (int i = 0; i < s1.length(); i++){
            char curr = s1.charAt(i);
            freqMap.put(curr, freqMap.getOrDefault(curr, 0) + 1);
        }
        
        int windowStart = 0, matched = 0;
        for (int windowEnd = 0; windowEnd < s2.length(); windowEnd++){
            char curr = s2.charAt(windowEnd);
            if (freqMap.containsKey(curr)){
                freqMap.put(curr, freqMap.get(curr) - 1);
                if (freqMap.get(curr) == 0)
                    matched++;
            }
            
            if (matched == freqMap.size())
                return true;
        
            if (windowEnd -windowStart + 1 >= s1.length()){
                char left = s2.charAt(windowStart);
                windowStart++;
                if (freqMap.containsKey(left)){
                    if (freqMap.get(left) == 0)
                        matched--;
                    freqMap.put(left, freqMap.get(left) + 1);
                }
                
            }
        }
        
        return false;
    }
    
}

File: 567-permutation-in-string/567-permutation-in-string.py
class Solution:
    def checkInclusion(self, s1: str, s2: str) -> bool:
        char_freq_map = {}
        
        for char in s1:
            if char not in char_freq_map:
                char_freq_map[char] = 0
            char_freq_map[char] += 1
        window_start = 0
        matched = 0
        for window_end in range(len(s2)):
            right_char = s2[window_end]
            
            if right_char in char_freq_map:
                char_freq_map[right_char] -= 1
                if char_freq_map[right_char] == 0:
                    matched += 1
            
            if matched == len(char_freq_map):
                return True
            
            if (window_end - window_start  + 1 >= len(s1)):
                left_char = s2[window_start]
                if left_char in char_freq_map:
                    if char_freq_map[left_char] == 0:
                        matched -= 1
                    char_freq_map[left_char] += 1
                window_start += 1
            
            
        return False